fix(client-presence): [BREAKING CHANGE] `LatestMap` keys limited to `string`s

jason-ha opened the pull request:
Remove `number` as a possible type for `LatestMap` keys. All given keys are stored as string record entries at runtime and thus numbers become keys. To use "number" keys, string pattern `${number}` can be used for the type.

Add `keyValidator` to support enabling consumers to only work with validated keys. A `keyValidator` is recommended anytime `Keys` type is not `string`.

jason-ha committed:
feat(client-presence): LatestMap key validation option
A `keyValidator` may now be given when creating a `LatestMap` such that only validated keys are enumerated.

tylerbutler reviewed (approved):
If I understand correctly, there's no way for a user to access or enumerate keys for which the validator returns undefined. I can understand this behavior for iteration, but it seems odd that .get(invalidKey) is the same as .get(missingKey). In practice maybe that distinction doesn't matter much.

jason-ha committed:
improvement(client-presence): refine key validator

jason-ha committed:
fix(client-presence): retain all remote updates even with invalid keys

jason-ha committed:
fix(client-presence): [BREAKING CHANGE]: remove `number` key from `LatestMap`
`number`s have always propagated as `string`s at runtime. Removal of specification makes API types reflect this.

Internally, `objectEntries` appears to fail to fully preserve key enumeration not that `number` is not possible.

jason-ha committed:
docs(client-presence): internal ts-doc and generic name improvements

jason-ha committed:
fix(client-presence): no remoteUpdated when all keys are invalid
add key validator test coverage

jason-ha committed:
test(client-end-to-end): lint fixups

jason-ha committed:
merge: 'main' into presence/map-key-validator

tylerbutler reviewed (approved):
LGTM.

jason-ha committed:
merge: 'main' into presence/map-key-validator

jason-ha committed:
docs(client-presence): correct issue reference typo

jason-ha committed:
docs(client-presence): update alpha API report

jason-ha committed:
docs(client-presence): fix comment and test description typos

Josmithr commented on packages/framework/presence/src/internalTypes.ts:
```diff
@@ -79,6 +79,13 @@ export interface ValueManager<
 	readonly value?: TValueState;
+
+	/**
+	 * Process an update of value for remote attendee.
```
Nit (to make explicit that we're referring to a specific entity and not a concept):
```suggestion
	 * Process an update of `value` for remote attendee.
```

jason-ha commented on .changeset/modern-candies-clean.md:
```diff
@@ -0,0 +1,7 @@
+---
+Removal of number key support in LatestMap
+
+`number` keys have never been successfully propagated as `number`s at runtime and this type clarification makes that clear. See [issue 25919](https://github.com/microsoft/FluidFramework/issues/25919) for more details.
```
👍Issue updated with that workaround.

jason-ha committed:
merge: 'main' into presence/map-key-validator

jason-ha committed:
docs(client-presence): updated comments
